Scripts evaluation build operations

eskatos opened the pull request:
Introduce build operations around init, settings, build, plugins script evaluation.

- [ ] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior.

bamboo commented on subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultObjectConfigurationAction.java:
```diff
@@ -97,9 +102,14 @@ private void applyScript(Object script) {
-        ScriptPlugin configurer = configurerFactory.create(scriptSource, scriptHandler, classLoaderScopeChild, classLoaderScope, false);
-        for (Object target : targets) {
-            configurer.apply(target);
+        final ScriptPlugin configurer = configurerFactory.create(scriptSource, scriptHandler, classLoaderScopeChild, classLoaderScope, false);
```
Shouldn't the call to `create` also be part of the build operation?

In the case of `gradle-script-kotlin` that would account for the time spent resolving and instantiating service dependencies.

eskatos replied:
Yeah I thought about that. My rationale for this first iteration is that only the actual 'application' of the scripts is taken into account. But we might want to include classloader scope, script handler and script plugin creations.

In this particular case we are applying to a set of targets. We need a build operation for each, only the first one would then account for the setup.

@adammurdoch, do you concur?

bamboo commented on subprojects/core/src/main/java/org/gradle/configuration/DefaultInitScriptProcessor.java:
```diff
@@ -32,20 +35,27 @@
         ScriptHandler scriptHandler = scriptHandlerFactory.create(initScript, scriptScope);
-        ScriptPlugin configurer = configurerFactory.create(initScript, scriptHandler, scriptScope, baseScope, true);
-        configurer.apply(gradle);
+        final ScriptPlugin configurer = configurerFactory.create(initScript, scriptHandler, scriptScope, baseScope, true);
```
Didn't I just see this code somewhere? ;)

adammurdoch replied:
it should be inside the call to `apply()`

bamboo commented on subprojects/core/src/main/java/org/gradle/configuration/project/BuildScriptProcessor.java:
```diff
@@ -26,17 +29,24 @@
         try {
-            ScriptPlugin configurer = configurerFactory.create(project.getBuildScriptSource(), project.getBuildscript(), project.getClassLoaderScope(), project.getBaseClassLoaderScope(), true);
-            configurer.apply(project);
+            final ScriptPlugin configurer = configurerFactory.create(project.getBuildScriptSource(), project.getBuildscript(), project.getClassLoaderScope(), project.getBaseClassLoaderScope(), true);
```
Same as above.

bamboo commented on subprojects/core/src/main/java/org/gradle/initialization/ScriptEvaluatingSettingsProcessor.java:
```diff
@@ -69,8 +75,12 @@ private void applySettingsScript(SettingsLocation settingsLocation, final Settin
         ScriptHandler scriptHandler = scriptHandlerFactory.create(settingsScriptSource, settingsClassLoaderScope);
-        ScriptPlugin configurer = configurerFactory.create(settingsScriptSource, scriptHandler, settingsClassLoaderScope, settings.getRootClassLoaderScope(), true);
-        configurer.apply(settings);
+        final ScriptPlugin configurer = configurerFactory.create(settingsScriptSource, scriptHandler, settingsClassLoaderScope, settings.getRootClassLoaderScope(), true);
```
Same as above.

bamboo commented on subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ProgressEvents.groovy:
```diff
@@ -282,32 +282,51 @@ class ProgressEvents implements ProgressListener {
             def found = [] as List<Operation>
-            def descendantsText = ''
             def recurse
-            recurse = { List<Operation> children, int level = 0 ->
```
Join variable declaration and assignment?

eskatos replied:
Nope, if joined, then the reference to `recurse` inside the closure doesn't compile.

bamboo replied:
I see.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultObjectConfigurationAction.java:
```diff
@@ -97,9 +102,14 @@ private void applyScript(Object script) {
-            configurer.apply(target);
+        final ScriptPlugin configurer = configurerFactory.create(scriptSource, scriptHandler, classLoaderScopeChild, classLoaderScope, false);
+        for (final Object target : targets) {
+            buildOperationExecutor.run("Apply " + scriptSource.getDisplayName() + " to " + target, new Action<BuildOperationContext>() {
```
this should be inside the `apply()` method, as this should just be a fundamental part of applying a script plugin, regardless of where and to what we're applying the script.

adammurdoch commented on subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r33/BuildProgressCrossVersionSpec.groovy:
```diff
@@ -153,19 +153,10 @@ class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
+        events.operation("Configure build")
+            .descendant("Configure project :")
+            .descendant("Configure project :a")
+            .descendant("Configure project :b")
```
these shouldn't be changing

adammurdoch commented on subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r35/BuildProgressCrossVersionSpec.groovy:
```diff
@@ -67,31 +67,18 @@ class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
         then:
         events.assertIsABuild()
 
-        def configureBuild = events.operation("Configure build")
```
it's important that the new events appear in the correct location, ie children of the relevant configure project event, so we should test for this.

eskatos replied:
done

eskatos commented:
About the boundaries of the added build operations.

My changes above put them around each call to `ScriptPlugin#apply(target)`.

`ScriptPlugin` instances are created by `ScriptPluginFactory` instances that can come from different providers (Groovy, Kotlin).

If I move the added build operations boundaries inside `DefaultScriptPlugin`, we'll have to do it properly too inside each provider implementation.

I get your point @adammurdoch, semantically the build operation should be part of `ScriptPlugin#apply(target)`. This could be done without putting this burden on providers by wrapping `ScriptPluginFactory` instances. But, it won't take into account some of the processing needed to actually "apply a script".

I also get @bamboo's point in that we should include `ScriptPluginFactory#create(..)` into these build operations. This would account for `ScriptPluginFactorySelector` processing and would gives us a way to fairly compare providers.

In any case, `ScriptPlugin#apply(target)` does a lot more than just "applying the script", there is multi-phase compilation, caching, multi-phase application etc...

So, we could add more build operations to distinguish more steps. But they're not free.
My take would be to stick to the build ops added above and include `ScriptPluginFactory#create(..)`.

@adammurdoch, @bamboo, wdyt?

eskatos commented:
I rebased the branch on `master` and introduced `ScriptPluginApplicator` that includes both `ScriptPluginFactory#create(..)` and `ScriptPlugin#apply(target)`, and, emits build operations events around scripts applications. Build operations related code is not spread all over the place anymore.

This should comply with the concerns of you two, @adammurdoch and @bamboo

jjohannes commented on subprojects/core/src/main/java/org/gradle/api/internal/configuration/DefaultScriptPluginApplicator.java:
```diff
@@ -0,0 +1,78 @@
+            @Override
+            public void execute(BuildOperationContext buildOperationContext) {
+                ScriptPlugin scriptPlugin = scriptPluginFactory.create(scriptSource, scriptHandler, targetScope, baseScope, topLevelScript);
+                scriptPlugin.apply(firstTarget);
```
Maybe we could change `operationDisplayNamePrefix ` for the first application (which includes `create()`) to `Create and apply...`

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/configuration/DefaultScriptPluginApplicator.java:
```diff
@@ -0,0 +1,78 @@
+        final Object firstTarget = it.next();
+        final List<ScriptPlugin> scriptPluginHolder = new ArrayList<ScriptPlugin>(1);
+        String operationDisplayNamePrefix = "Apply " + scriptSource.getDisplayName() + " to ";
+        buildOperationExecutor.run(operationDisplayNamePrefix + firstTarget, new Action<BuildOperationContext>() {
```
I think we're trying to solve a problem that doesn't really exist, or at least a different problem. I think we just want to decorate the call to `apply()`, whose contract is "do whatever you need to do apply this script to this target". This is where the user code happens. Let's move `create()` out. This is, or should be, cheap and there's no user code that happens during this.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/configuration/DefaultScriptPluginApplicator.java:
```diff
@@ -0,0 +1,78 @@
+        if (target instanceof Collection) {
+            targets = Cast.uncheckedCast(target);
+        } else {
+            targets = Collections.singleton(target);
```
we should strongly type this, i.e. make the parameter `List<?> targets` rather than `Object target`.

adammurdoch commented on subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java:
```diff
@@ -124,7 +130,7 @@ public void apply(final Object target) {
                 }
             };
-            services.add(ScriptPluginFactory.class, DefaultScriptPluginFactory.this);
+            services.add(ScriptPluginApplicator.class, new DefaultScriptPluginApplicator(DefaultScriptPluginFactory.this, buildOperationExecutor));
```
not really related to this change, but adding this service means that a `.gradle` script will only be able to apply other `.gradle` scripts, rather than a script of any type supported by Gradle.

oehme replied:
I think this was the case before as well

adammurdoch replied:
that's right. we might fix it given we're restructuring the services

adammurdoch commented on subprojects/core/src/main/java/org/gradle/configuration/project/BuildScriptProcessor.java:
```diff
@@ -15,28 +15,26 @@
 
-    public void execute(ProjectInternal project) {
+    public void execute(final ProjectInternal project) {
         LOGGER.info("Evaluating {} using {}.", project, project.getBuildScriptSource().getDisplayName());
```
let's get rid of this logging, as wrapping script execution in a build operation will provide this kind of data anyway.

jjohannes commented:
As discussed with @oehme, I have reworked the solution to wrap the ScriptPlugin implementations inside a  decorator `BuildOperationScriptPlugin`. This creates the build operation inside the `apply()`.
https://github.com/gradle/gradle/pull/1527/commits/7c3695712078392fe55ab9918d63506c952f41ee

To make this work for nested scripts, I had to fix the problem that the `DefaultScriptPluginFactory` was only using itself for nested applications. I think that's a general fix we want (as already suggested in one earlier comment by @adammurdoch). This is not very beautiful and binds `DefaultScriptPluginFactory` and `ScriptPluginFactorySelector` very strongly together. But it was the best _simple_ solution I came up with.
https://github.com/gradle/gradle/pull/1527/commits/4be6c9443c4a538df40f7c52fa5cddb505fd3658

adammurdoch commented on subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r36/BuildProgressCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,304 @@
+
+@ToolingApiVersion(">=3.6")
+@TargetGradleVersion(">=3.6")
+class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
```
Given the awkward wiring, it might be good to add some basic coverage that we get some events for kotlin build scripts.

jjohannes commented:
I added a bunch of test coverage. Running CI branch build now. If that passes, I will merge.

https://builds.gradle.org/viewType.html?buildTypeId=Gradle_Branches_Checkpoints_Stage5Full&branch_Gradle_Branches_Checkpoints=pm-build-scripts-build-ops&tab=buildTypeStatusDiv

oehme commented on subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java:
```diff
@@ -287,8 +287,11 @@ public boolean isValid() {
+    protected ScriptPluginFactory createScriptPluginFactory(BuildOperationExecutor buildOperationExecutor) {
+        DefaultScriptPluginFactory defaultScriptPluginFactory = defaultScriptPluginFactory();
+        ScriptPluginFactorySelector scriptPluginFactorySelector = new ScriptPluginFactorySelector(defaultScriptPluginFactory, this, buildOperationExecutor);
+        defaultScriptPluginFactory.setScriptPluginFactory(scriptPluginFactorySelector);
```
Should probably be a constructor argument.

jjohannes replied:
Like this it is not possible, because we construct two objects that reference each other in both directions. Only one can have the other as constructor argument.

oehme commented on subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java:
```diff
@@ -124,7 +130,7 @@ public void apply(final Object target) {
                 }
             };
-            services.add(ScriptPluginFactory.class, DefaultScriptPluginFactory.this);
+            services.add(ScriptPluginFactory.class, scriptPluginFactory);
```
One way to detangle this would be to pass this service registry from the outside instead of building it here. Would also reduce the number of constructor arguments.

jjohannes replied:
I will give it a try.

jjohannes replied:
The service registry is specific for each `ScriptPluginImpl`. That's why we build it here. Each instance has a potentially different `ClassLoaderScope`.

oehme replied:
I think you can completely break the cycle by making the selector not depend on the DefaultScriptPluginFactory. Instead, the DefaultScriptPluginFactory should live in the `pluginUse` project and be discovered using the same ServiceLocator pattern that the KotlinScriptPluginFactory uses.

Let's not keep this PR waiting, we can do that in a follow-up.

adammurdoch commented on subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/r36/BuildProgressCrossVersionSpec.groovy:
```diff
@@ -73,4 +83,311 @@ class BuildProgressCrossVersionSpec extends ToolingApiSpecification {
         events.operation('Task :runForked').descendant('My forked worker action')
     }
+
+    @ToolingApiVersion(">=3.6")
```
Should be `@TargetGradleVersion`, there's no changes to the tooling API. The test should describe which versions of Gradle and the tooling API that the feature will work with, not which versions it should be tested against - that's decided by something else.

Same for the other methods of this test class.

jjohannes replied:
You are right. I got that mixed up here. I removed the limitations.

eskatos commented:
🎉
